### PR TITLE
[Proposal] switch to furo theme

### PR DESCRIPTION
**Pull request by 2bndy5**
Following the progress from the post-merging dialogue in #759

- add logo for branding in favicon and nav menu title
- remove all `.. contents::` directives as they're redundant in the furo theme (ToC has an independent menu)
- uniform some theme elements' colors (via custom CSS) to match the blue in breathe-doc.org
- add the [sphinx-copybutton](https://github.com/executablebooks/sphinx-copybutton#readme) extension. This popular ext from [executablebooks](https://github.com/executablebooks) automatically adds a "copy to clipboard" button to all code blocks (including those from C/C++ sources).

### CODE REVIEW

**Review by vermeeren — APPROVED** [review]
@2bndy5 Nice stuff!

@michaeljones You also ok with this?

**Comment by 2bndy5** [comment]
Thanks for the logo improvements. I'm not proficient with inkscape as I'd like to be.

**Comment by 2bndy5** [comment]
I can re-make the favicon from the revised logo (takes 1 minute in GIMP).

**Comment by michaeljones** [comment]
All good. Do you think the changes are ok?

**Comment by 2bndy5** [comment]
I think we don't need to explicitly control the color of the nav menu title now that the logo's bg color is the theme default.

**Comment by 2bndy5** [comment]
ok, now I'm good to go here

**Comment by 2bndy5** [comment]
> Do you think the changes are ok? Overall?

Yes, I like the text's blue better than the theme's default. Good call.

FYI, I did try to experiment getting the main content BG color to match the body BG color of breathe-doc,org, but it messed up the ToC coloring and made it surprisingly more difficult to do. Thus, I just stuck to changing the mobile header's color.

**Comment by michaeljones** [comment]
I've adjusted it a bit more to work better against the less contrasty side bar background.

I'm happy to merge now I think. If you are?

**Comment by 2bndy5** [comment]
yep. lgtm

**Comment by jakobandersen** [comment]
Does the furo theme derive from the Sphinx basic theme?
There are some recent'ish style additions in the basic theme to syntax highlight C and C++ declarations, that I think should we should use (in adapted form): https://github.com/sphinx-doc/sphinx/blob/b3812f72a98b01bae4b1158761082edc46cfa87f/sphinx/themes/basic/static/basic.css_t#L546

**Comment by jakobandersen** [comment]
Actually, the full signature CSS stuff starts at https://github.com/sphinx-doc/sphinx/blob/b3812f72a98b01bae4b1158761082edc46cfa87f/sphinx/themes/basic/static/basic.css_t#L511

**Comment by 2bndy5** [comment]
Yes, [furo inherits from basic CSS](https://github.com/pradyunsg/furo/blob/641f14d08c460aa691593e784f1d4dee51232825/src/furo/theme/furo/theme.conf#L2)

**Comment by 2bndy5** [comment]
> There are some recent'ish style additions in the basic theme to syntax highlight C and C++ declarations, that I think should we should use (in adapted form)

@jakobandersen How "adapted" were you thinking?

**Comment by jakobandersen** [comment]
> Yes, [furo inherits from basic CSS](https://github.com/pradyunsg/furo/blob/641f14d08c460aa691593e784f1d4dee51232825/src/furo/theme/furo/theme.conf#L2)

Hmm, indeed, though the basic CSS does not seem applied in the output. In my built ``_static/`` folder there is ``basic.css``, but it is not included in the HTML. I'm not sure how the exact mechanism for theme derivation actually is supposed to work though.

**Comment by jakobandersen** [comment]
> > There are some recent'ish style additions in the basic theme to syntax highlight C and C++ declarations, that I think should we should use (in adapted form)
> 
> @jakobandersen How "adapted" were you thinking?

I selected the colours in the basic theme as a, hopefully, good default. So I was mostly thinking about adapting the colours to fit in with the rest of the theme.

**Comment by jakobandersen** [comment]
> > Yes, [furo inherits from basic CSS](https://github.com/pradyunsg/furo/blob/641f14d08c460aa691593e784f1d4dee51232825/src/furo/theme/furo/theme.conf#L2)
> 
> Hmm, indeed, though the basic CSS does not seem applied in the output. In my built `_static/` folder there is `basic.css`, but it is not included in the HTML. I'm not sure how the exact mechanism for theme derivation actually is supposed to work though.

Ah, from https://www.sphinx-doc.org/en/master/development/theming.html:
> The inherit setting [...]. If you want to also inherit the stylesheet, include it via CSS’ @import in your own.

I guess this might not be done? I have noticed a similar problem with the RTD theme which also lacks the new syntax highlighting.

**Comment by 2bndy5** [comment]
With the support for light and dark theme, I wouldn't be surprised if furo overrides the colors used in the signatures.

Generally the CSS is included from the [base.html template](https://github.com/pradyunsg/furo/blob/641f14d08c460aa691593e784f1d4dee51232825/src/furo/theme/furo/base.html#L57-L72), but furo uses some other partial templates to include a multitude of CSS (there's currently a special CSS file for the sphinx-design ext that will be going away in the future).

Looking at the [furo python src](https://github.com/pradyunsg/furo/blob/641f14d08c460aa691593e784f1d4dee51232825/src/furo/__init__.py#L354-L365), the pygments CSS is generated for the theme... I don't see a use of the _basic.css_ anywhere. Maybe the theme src was just following the typical theme development at first, then it had a major design decision about coloring?

Is this cause to reject the proposal? Did you want to stick with haiku or maybe some other theme that brings in your changes to basic..css? IMO, the haiku theme very dated in terms of material design. I don't like the RTD theme anymore after seeing how poorly compatible with mobile displays (terrible for long pages).

**Comment by jakobandersen** [comment]
> Is this cause to reject the proposal? Did you want to stick with haiku or maybe some other theme that brings in your changes to basic..css? IMO, the haiku theme very dated in terms of material design. I don't like the RTD theme anymore after seeing how poorly compatible it is with mobile displays (terrible for long pages).

Not at all, I think the new one looks great, except for the lack of syntax highlighting. I find C, and in particular, C++ declarations difficult to read without it.
I suggest we for now copy-paste relevant parts form the basic theme into https://github.com/michaeljones/breathe/blob/master/documentation/source/_static/breathe.css.
I just tried locally to copy https://github.com/sphinx-doc/sphinx/blob/b3812f72a98b01bae4b1158761082edc46cfa87f/sphinx/themes/basic/static/basic.css_t#L511 through line 568. The change to a monospace font makes it larger, but adding ``font-size: 0.9em;`` to ``.sig`` seems to makes it appear similar to the old one in size.
But I'm really no expert in CSS so overall my goal is: use monospace font and syntax highlighting for signatures, so reading documentation is not significantly different from scrolling through ordinary code.

**Comment by 2bndy5** [comment]
The font-size for a signatures use a CSS variable `--font-size--small: 87.5%;` Not sure if this helps, but you could try
```css
font-size: var(--font-size--small);
```

The bulk of the furo CSS is compiled from SCSS (which I only recently learned to use). The theme's docs do outline some [guidance about customizing the colors](https://pradyunsg.me/furo/customisation/colors/#).

**Comment by 2bndy5** [comment]
Looking at the basic.css (lines only in the section that was linked)
1. I'm not too keen on seeing italics in code snippets.
2. The colors for the basic.css are hard coded. This meabs they will not change depending on the theme's color scheme (light vs dark).
3. using `font-family: unset;` generally assumes that the inherited value is undesirable. This is _very_ theme dependent.

> my goal is: use monospace font and syntax highlighting for signatures

The font-family is already using a monospace font. Specifically (in order of fall back),
```css
--font-family--monospace: "SFMono-Regular",Menlo,Consolas,Monaco,Liberation Mono,Lucida Console,monospace;
```
Also, the basic.css doesn't seem to offer much more syntax highlighting than what I already see in furo. I did notice the parameters' names aren't colored like they'd be in an IDE. But that is only 1 new CSS rule away from fixing it.
```css
.sig .n {
  /* set your color for light and dark theme with a variable */
  color: var(--breathe-param-color);
}
```

**Comment by 2bndy5** [comment]
Ok, I added some syntax highlighting that is color scheme dependent. I also de-saturated that ugly yellow background that most themes use for code snippets (🤮 its so hard to see syntax highlighted colors when the BG color is noticeably saturated).

**Comment by michaeljones** [comment]
Sorry, I've had a busy few weeks. Where are we at with this? Any outstanding issues? Are we in a place to merge?

**Comment by 2bndy5** [comment]
@jakobandersen voiced concerns about syntax highlighting in signatures, so I made some additional CSS alterations. Still waiting to hear if those newer CSS mods are satisfactory.

**Comment by jakobandersen** [comment]
> @jakobandersen voiced concerns about syntax highlighting in signatures, so I made some additional CSS alterations. Still waiting to hear if those newer CSS mods are satisfactory.

I gave it a quick glance now and it looks quite good to me. Though, one thing I suggest changing is to make declared names and links more distinct. Example, in the signature ``void crack(Tool tool)``, the name ``crack`` is in bold blue, and the linked type ``Tool`` is in thinner blue. My non-designer gut-feeling would be to make the name bold black as it is the most prominent.

File: documentation/source/_static/breathe.css [hunk 1]
```diff
@@ -1,13 +1,104 @@
...
+  color: var(--breathe-number-color);
+}
+
+.sig.c   .s, .sig.c   .sc,
+.sig.cpp .s, .sig.cpp .sc {
+  color: var(--breathe-str-char-color);
 }
 
+.sig > .n {
+  color: var(--breathe-param-color);
```
Comment (jakobandersen) [inline documentation/source/_static/breathe.css:103]:
```suggestion
  color: var(--breathe-name-color);
```

Comment (jakobandersen) [reply documentation/source/_static/breathe.css:103]: To be more in line with the Sphinx naming and the semantic purpose of the ``n`` class.
(And similarly for all the variations of this variable)

**Review by jakobandersen — CHANGES_REQUESTED** [review]

**Comment by 2bndy5** [comment]
Bold black will work for the light theme, but I'll invert that to bold white for the dark theme.

I'm surprised the CSS var name you suggest is more preferred. It seems a bit non-descriptive. Still, I will change it as requested.

File: documentation/source/_static/breathe.css (hunk 1 above)
Comment (2bndy5) [reply documentation/source/_static/breathe.css]: done

**Comment by 2bndy5** [comment]
> Though, one thing I suggest changing is to make declared names and links more distinct. Example, in the signature void crack(Tool tool), the name crack is in bold blue, and the linked type Tool is in thinner blue. 

@jakobandersen I don't see this blue color used for `crack` in `void crack(Tool tool)`. With the latest builds, the name `crack` is in red. What page did you observe this blue coloration on? 

###  From the bottom of index.rst
Dark theme
image

Light theme
image

**Comment by michaeljones** [comment]
Having asked, I'm afraid I'm now going to be quite busy for a bit. I'm happy for this to be merged when you both are. Otherwise I'll try to reconnect with it in a few weeks.

**Comment by jakobandersen** [comment]
> @jakobandersen I don't see this blue color used for `crack` in `void crack(Tool tool)`. With the latest builds, the name `crack` is in red. What page did you observe this blue coloration on?

I think it was because I had an older version of the furo theme installed. Once I updated it I see the same colours as in your screen dump.

A small thing I randomly noticed is on ``autoindex.html``. Here the class ``AutoClassTest`` and its members have font weight 700 (which seems correct from ``.sig``). But below, the free functions ``autoFunction`` and ``anotherAutoFunction`` gets an overwritten font weight of 500. Perhaps this is some ickyness of the doctree generation from Breathe, but maybe it's a CSS issue.

**Comment by jakobandersen** [comment]
There is some indentation ickyness for multi-line signatures. E.g., in ``class.html``, the class template ``TemplateClass``, the first line (the template parameter list) is indented with ``text-indent: -2.5em;`` but not the following line with the class declaration.

**Comment by 2bndy5** [comment]
I think you found a problem with the theme CSS (not any of our custom CSS). It seems that the -2.5em indent is trying to compensate for an unusual padding-left of 3em (also set in the `.sig` rule). I'll raise this concern in the furo theme repo and see if the author is open to addressing it there.

Since this PR isn't going anywhere anytime soon (& there's no rush), I'll update this thread/branch when appropriate,

**Comment by 2bndy5** [comment]
Opened a "potential bug" discussion over at the furo theme repo: https://github.com/pradyunsg/furo/discussions/427

**Comment by 2bndy5** [comment]
@jakobandersen I pushed a CSS bugfix that resolves the multi-lined improper indent problem. Hopefully it will get addressed upstream in the furo theme's repo, but I haven't gotten any response there about it. Let me know if you spot anything else.

**Comment by michaeljones** [comment]
Now we have some conflicts here. I can look at them before too long, I imagine. 

I think I'm in favour of merging this too and treating any potential issues as further work.

It is exciting to have a fresh look.

**Comment by 2bndy5** [comment]
I've resolved the conflicts. Last I checked the `breatheenumvalues` and `breatheparameterlist` selectors in the custom CSS had no real impact (don't think they are actually getting used). I've been leaving them in for posterity since the docs seem to suggest doing so.

**Comment by vermeeren** [comment]
@2bndy5 @michaeljones If I understand correctly this is ready for merging (minus a rebase and perhaps some commit message rewording)? Diff looks good to me and that's a very good looking dark theme as well!

> I think I'm in favour of merging this too and treating any potential issues as further work.

Agreed. If I can get confirmation this is done (future things future issues) I'll rebase+merge. Feel free to also rebase @2bndy5 if you feel like it. Thanks!

**Comment by 2bndy5** [comment]
I did merge in the changes from master since the doc overhaul PR was merged. @vermeeren If you want a different git history (with better word choice in the messages), than I will leave that to you. I'm not against another squashing (which is probably preferable since there's a merge commit in this history).

**Comment by 2bndy5** [comment]
Is there an unwritten convention about acceptable commit messages for this project?

**Comment by vermeeren** [comment]
> Is there an unwritten convention about acceptable commit messages for this project?

@2bndy5 For Breathe there is nothing defined. Generally speaking for any project a commit message should be short and concise for the summary with, if needed, additional explanation and rationale further in the commit body. (Basically the so-called 50/72 rule, though the 50 is a guideline and not a hard cap.)

Consider someone is reading through the commit history for whatever reason (how a feature came to be, finding where there is a regression, understand/learning components, etc), this someone has no context of the PR and/or related issue in question. So a message summary such as `color touch ups` raises some questions. It should if possible satisfy the "what? where? of a commit, with the extended body doing the "why? how?" part.

So if someone with no knowledge of the actual code, who is new/unrelated to Breathe, but who does know the subject matter, reads just the `git log --oneline`, `git shortlog`, etc would that the coherent? I try to write for this future audience, possibly a future contributor, and not for myself or the people involved with the PR/issue. Maybe someone `git blame`s a "weird" line of code in 5 years and then finds a commit explaining why some edge case this and that with refs instead of `fix bug` as the entire message.

I think one of the best examples to learn from is the kernel commit style, see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/ This is also my preferred style. However anything is fine as long as it it clear and concise, as Breathe has no formal format defined.

Example: For this specific PR, the "what?" is the theme. So prefixing like `doc theme:` (or similar) would already massively improve with the log coherency. Compare `add syntax highlighting` with `doc theme: add syntax highlighting`. The first could be many things in Breathe, the second is instantly obvious. Writing good messages is actually quite hard and somewhat of an art form just like clean/simple code is.

P.S. Ended up very long, hopefully it's of use. :)

**Comment by 2bndy5** [comment]
A long comment is better than a short one, in my book. Once upon a time, I did start reading the Linux kernel project's contribution guidelines, but I stopped when I hit the third "biased" and extremely opinionated (more like racist) comments in the doc.

**Comment by vermeeren** [comment]
@2bndy5 I haven't read their guidelines or contributing docs (I think I almost never do for any project, instead mimicking recent commits' style), so only have experience with the practical side of things.

**Comment by michaeljones** [comment]
@vermeeren - I'm happy to see this merged. Do you want to do some rebasing and stuff on it and merge when you're happy? Otherwise we can just merge as is. I'm keen on a clean history but it is more important to get the actual work in especially as it has been open for so long now.

**Comment by vermeeren** [comment]
The readthedocs build is failing https://readthedocs.org/projects/breathe/builds/17362413/ saying `sphinx_copybutton` is not installed. I find this odd as our `.readthedocs.yaml` clearly specifies:

```yaml
python:
  version: 3
  install:
    - requirements: requirements/development.txt
    - method: pip
      path: .
```

And this PR does correctly add the new deps to that requirements file.

I do see caba0619107eb89886a9166a3fbd8ca2e3b0e978 added some `conda` settings, reading RTD yaml docs https://docs.readthedocs.io/en/latest/config-file/v2.html#requirements-file there is a warning that if using a Conda environment this setting will have no effect.

I checked https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html and found out how to use conda instead and still use the requirements text file. Test pushed this to a branch and made RTD build this, worked fine.

Pushed to master directly as f1e22cb871b410700cf6b1004363d3ceedacfb6a.

**Comment by vermeeren** [comment]
@michaeljones Could you check the project settings and have a look at that? Would be very useful indeed. (I think only the owner can access for user namespace projects.)

**Pull request merged**